src/recvsend.rs

```diff
@@ -338,15 +343,19 @@ where
         futures_util::ready!(self.framed.poll_ready_unpin(cx).map_err(Error::Send))
             .map_err(into_io_error)?;
 
-        let mut guard = self.shared.lock().unwrap();
-
-        let response = match guard.message.take() {
-            Some(response) => response,
-            None => {
-                guard.waker = Some(cx.waker().clone());
-                drop(guard);
-
-                return Poll::Pending;
+        let response = {
+            let mut guard = self.shared.lock().unwrap();
+
+            match guard.message.take() {
+                Some(response) => response,
+                None => {
+                    return if guard.waker.replace(cx.waker().clone()).is_none() {
```

Comment:
Does this work correctly?

Don't we always start with a waker being set now? (line 212)
We always take the waker out on drop.

In what circumstance would the waker here be `None`?

I think a boolean flag `dropped` or `canceled` would make this a bit easier to understand. Maybe introduce a function `is_canceled` on `Shared`?

Reply:
With these changes waker will be `None` if responder is simply dropped without sending any value.

The idea behind these changes is to intentionally support requests that don't have responses (like `AddProvider` in `libp2p-kad`). Since waker calling in this PR is moved to `Drop` impl, it will be always called, even when `Responder::respond()` isn't.

I think this addresses all 3 comments you had here.

Reply:
> With these changes waker will be `None` if responder is simply dropped without sending any value.

The waker being `None` here seems unrelated to me to the message being `None` or `Some`?

Isn't it the case that if we get in here and the waker is still set, it must mean that the `Responder` has not been used yet at all and is still alive?

I think it would be easier to understand if we'd do the following at the top:

```rust
if guard.waker.is_some() {
    // spurious wake-up
    // `Responder` wakes on drop so if we get here, it is still alive, just update the waker and continue waiting
    guard.waker = Some(cx.waker.clone());
    return Poll::Pending
}
```

> The idea behind these changes is to intentionally support requests that don't have responses (like `AddProvider` in `libp2p-kad`).

We will definitely have to support this but I am not sure this is the best way to do so. It might be better to create a newtype `AddProviderResponse` that is empty and acts as a token that allows you to control, when you are done with your `AddProvider` message. The newtype would map to an empty message in the `Encoder`, i.e. nothing being sent.

I think not having a message here should be an error. These silent failures are quite hard to debug otherwise.

Reply:
> > The idea behind these changes is to intentionally support requests that don't have responses (like `AddProvider` in `libp2p-kad`).
> 
> We will definitely have to support this but I am not sure this is the best way to do so. It might be better to create a newtype `AddProviderResponse` that is empty and acts as a token that allows you to control, when you are done with your `AddProvider` message. The newtype would map to an empty message in the `Encoder`, i.e. nothing being sent.

If this doesn't work, we can make the behaviour configurable so in `libp2p-kad` we don't error for this particular case.

Reply:
> The waker being None here seems unrelated to me to the message being None or Some?
>
> Isn't it the case that if we get in here and the waker is still set, it must mean that the Responder has not been used yet at all and is still alive?

I changed the code to always set waker on creation. Since the only case when waker is `None` would be when responder is dropped, we this check indicates nothing will be able to send the message later, if both message and waker are `None` we're done processing, message can't possibly be set later. I can change it to a different if/else like you've described, but `.replace()` was kind of created for the purpose of replacing and checking previously stored value at the same time.

> I think not having a message here should be an error. These silent failures are quite hard to debug otherwise.

We can mark the struct with `#[must_use]`, then unused value will generate a compiler warning.

Wouldn't `AddProviderResponse` still generate a message with empty payload though? Seems a like a tricky workaround and since we can't force the code to send a message in Rust (you can still drop the value anyway, compile time warning is the best we can do without linear types).

We can also add explicit method in addition to `respond()` that will drop responder without sending a message to make it more readable/obvious.

Also we can make the message `Option<Response>`, signifying that responder can be dropped without sending a message, then receiving side can generate a warning, but I think it will still generate some message on the wire and here we want no message at all.

Reply:
> > The waker being None here seems unrelated to me to the message being None or Some?
> >
> > Isn't it the case that if we get in here and the waker is still set, it must mean that the Responder has not been used yet at all and is still alive?
> 
> I changed the code to always set waker on creation. Since the only case when waker is `None` would be when responder is dropped

I think we are saying the same thing. Can we extract the `if` condition into a variable `is_spurious_wakeup`? I think that would make it a lot clearer.

> > I think not having a message here should be an error. These silent failures are quite hard to debug otherwise.
> 
> We can mark the struct with `#[must_use]`, then unused value will generate a compiler warning.

Good idea! `Responder` should def be `#[must_use]`.

> Wouldn't `AddProviderResponse` still generate a message with empty payload though?

That depends entirely on our `Encoder` implementation. I'll have to try to see if we can get it to work.

> Seems a like a tricky workaround

I'd consider it to be okay to have a tricky workaround for non-obvious behaviour. `RecvSend` should be designed around always sending a response. Not sending one is a bug IMO.

Putting the workaround in the `libp2p-kad` code is IMO more maintainable because it alerts you that something weird is going on there, not in the `RecvSend` code.

> We can also add explicit method in addition to `respond()` that will drop responder without sending a message to make it more readable/obvious.

I think that is a good idea. Something like `ignore` or `forget`?

Reply:
I think the cleanest solution here would be to not tie codec and `SendRecv` together, but rather compose them later. Then during composition phase it would be possible to decide where not having a value is an error or not.

Otherwise let's say it is an error to not send a response in some cases, what will you do if there is no response? Before this PR it will just hang forever, clearly suboptimal. We could make it panic, but that is hardly ideal either. Print an error to logs? What if user didn't configure logger and doesn't see the error. We need some kind of explicit handling for this and it seems use case specific to me.

Reply:
> I think the cleanest solution here would be to not tie codec and `SendRecv` together, but rather compose them later. Then during composition phase it would be possible to decide where not having a value is an error or not.

How would that work? Relying on the codec is what makes this work, doesn't it?

> Otherwise let's say it is an error to not send a response in some cases, what will you do if there is no response?

Then the user should not use the `RecvSend` message pattern. My vision was to provide multiple of these state machines: `Send` (for identify push e.g.), `Recv` and `SendRecv`.

That of course is a slight problem for the `AddProvider` case because then the state machine doesn't wait for your processing. I wonder if we should perhaps extend the spec to include a response for `AddProvider` instead of hacking something together here or in `libp2p-kad`?

If we go for the "empty response hack", we would need to rewrite the `KadInStreamSink` to properly implement `Encoder` instead of using `with` composition and then simply write zero bytes to the buffer for the `AddProviderResponse`.

Reply:
Well, extending spec will probably make it incompatible with the current spec. Encoder hack feels wrong because to me `Some(&[])` and `None` are fundamentally different things.

Looks like this is controversial in the end. Feel free to pull pieces from this PR that you like or I can submit a PR with a subset of changes if you'd prefer without support for requests without responses.

Reply:
Let's merge as is. I'll follow-up with some commits in the PR! Thank you!